### PR TITLE
MIMO: add --deterministic flag to LLaVA(+audio) examples

### PR DESCRIPTION
**Pull request by kamran-nvidia**
# What does this PR do ?

MIMO: add --deterministic flag to LLaVA(+audio) examples

# Changelog

- Adds an opt-in `--deterministic` flag to the MIMO LLaVA and LLaVA-audio
  training scripts (`megatron_mimo_training_llava.py`,
  `megatron_mimo_training_llava_audio.py`). When set, vision, language, audio,
  and projection configs run in FP32 with unfused attention, loss fusion
  disabled, and full activation recompute so runs are bitwise reproducible.
- Threads the `deterministic` argument through the relevant config and model
  spec helpers so each sub-module is configured consistently.
- Updates the launcher shell scripts under
  `examples/models/megatron_mimo/` (`run_hetero_llava.sh`,
  `run_hetero_llava_audio.sh`, and the three `*_parallelism_tests*.sh`
  variants) to forward `--deterministic` and export the env vars required for
  deterministic execution (e.g. cuBLAS / NCCL settings).
- No behavior change when the flag is omitted — default training paths are
  untouched.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by kamran-nvidia** [comment]
/ok to test [1b54442](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3531/commits/1b54442efd8c9844dadbb42f1e398f4c6211ad09)

**Comment by kamran-nvidia** [comment]
/ok to test [4aafc26](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3531/commits/4aafc2614a6816b4dafa85445c898b5b41deb4e8)

**Comment by kamran-nvidia** [comment]
/ok to test [4c2f059](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3531/commits/4c2f059a4a7cced97a929e0f165eeaeceb13dbe4)

File: examples/models/megatron_mimo/megatron_mimo_training_llava.py [hunk 1]
```diff
@@ -75,12 +76,19 @@ def _make_vision_config() -> TransformerConfig:
     cfg.apply_rope_fusion = False
     # CLIP uses "quick_gelu", not standard gelu
     cfg.activation_func = lambda x: x * torch.sigmoid(1.702 * x)
-    cfg.calculate_per_token_loss = False
+    cfg.calculate_per_token_loss = True
+
+    if deterministic:
+        cfg.attention_backend = AttnBackend.unfused
+        cfg.deterministic_mode = True
+        cfg.recompute_granularity = "full"
```
Comment (yashaswikarnati) [inline examples/models/megatron_mimo/megatron_mimo_training_llava.py:84]: just a nit; do we actually need recompte flags for deterministic path ? like can we actually have no recompute?

File: examples/models/megatron_mimo/run_hetero_llava_audio_parallelism_tests.sh [hunk 2]
```diff
@@ -56,16 +62,50 @@ while [[ $# -gt 0 ]]; do
...
+if [[ "${DETERMINISTIC}" == "1" ]]; then
+    DETERMINISTIC_FLAG="--deterministic"
+    EXP_SUFFIX="-fp32"
+    CLIP_GRAD=${CLIP_GRAD:-0.0}
+    # Pin Ring algorithm for deterministic reduction order.
+    # Tree is faster for some message sizes but NCCL 2.28 Tree doesn't support
+    # AllGather with Int8 (used by torch.distributed.all_gather_object), and
+    # letting NCCL choose per-operation (^NVLS) still leaves Tree/Ring selection
+    # non-deterministic.  Ring supports all collective ops.
+    export NCCL_ALGO=Ring
```
Comment (yashaswikarnati) [inline examples/models/megatron_mimo/run_hetero_llava_audio_parallelism_tests.sh:91]: just for my own understanding, should this be tree or ring for more "deterministic" reduction?

**Comment by yashaswikarnati** [comment]
just some minor comments, lgtm overall! I also wonder if we should also have some of the shell scripts in the CI folder so tests are automatically run for these

File: examples/models/megatron_mimo/megatron_mimo_training_llava.py (hunk 1 above)
Comment (kamran-nvidia) [reply examples/models/megatron_mimo/megatron_mimo_training_llava.py:84]: Since we are switching the runs to FP32, most of the configs OOM with the same MBS as BFP16. Hence, I activated the activation recompute to fit the same MBS in both BFP16 and FP32.

File: examples/models/megatron_mimo/run_hetero_llava_audio_parallelism_tests.sh (hunk 2 above)
Comment (kamran-nvidia) [reply examples/models/megatron_mimo/run_hetero_llava_audio_parallelism_tests.sh:91]: Tree algorithm didn't work here since it didn't support AllGather with Int8 which was required somewhere in MIMO code. But from what I read, both algorithms (Tree and Ring) are deterministic in the sense that they allreduce in a known order. What causes non-determinism is the automatic switching between these two algorithms to optimizer for a specific architecture. So fixing to either should be fine.

File: examples/models/megatron_mimo/megatron_mimo_training_llava.py
Comment (liding-nv) [inline examples/models/megatron_mimo/megatron_mimo_training_llava.py]: i think those are ddp config, not trian config: `overlap_grad_reduce`, `use_distributed_optimizer`, `check_for_nan_in_grad`. 
and `log_interval` is in logger config

File: examples/models/megatron_mimo/run_hetero_llava_audio_parallelism_tests.sh (hunk 2 above)
Comment (yashaswikarnati) [reply examples/models/megatron_mimo/run_hetero_llava_audio_parallelism_tests.sh:91]: thanks for the info! not sure why there is a collective on int 8 tensors, may be some metadata exchange in somewhere in the e2e path..

**Review by yashaswikarnati — APPROVED** [review]

File: examples/models/megatron_mimo/megatron_mimo_training_llava.py
Comment (kamran-nvidia) [reply examples/models/megatron_mimo/megatron_mimo_training_llava.py]: @liding-nv Thanks for catching these. Looks like some AI slops that have been copied and carried around. Fixed here and in other files.

**Comment by kamran-nvidia** [comment]
@yashaswikarnati Added deterministic path to the functional tests

**Comment by kamran-nvidia** [comment]
/ok to test [86bafa3](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3531/commits/86bafa3b5afe6fbaafbc69699ad95b9ecedb24d8)

**Comment by kamran-nvidia** [comment]
/ok to test [a7c56fe](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3531/commits/a7c56fe2756cf1b28788d869b7c780567092fdf4)
